ql/src/java/org/apache/hadoop/hive/ql/udf/generic/AbstractGenericUDFArrayBase.java

```diff
@@ -0,0 +1,160 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.udf.generic;
+
+import org.apache.hadoop.hive.ql.exec.UDFArgumentException;
+import org.apache.hadoop.hive.ql.exec.UDFArgumentTypeException;
+import org.apache.hadoop.hive.serde.serdeConstants;
+
+import org.apache.hadoop.hive.serde2.objectinspector.ListObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorFactory;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorConverters;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorConverters.Converter;
+import org.apache.hadoop.hive.serde2.objectinspector.PrimitiveObjectInspector;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Abstract GenericUDF for array functions
+ */
+
+public abstract class AbstractGenericUDFArrayBase extends GenericUDF {
+
+    static final int ARRAY_IDX = 0;
+    static final int ARRAY2_IDX = 1;
+    static final int START_IDX = 1;
+    static final int LENGTH_IDX = 2;
+    static final int SEPARATOR_IDX = 1;
+    static final int REPLACE_NULL_IDX = 2;
+
+    int MIN_ARG_COUNT;
+    int MAX_ARG_COUNT;
+
+    transient ListObjectInspector arrayOI;
+    transient ObjectInspector[] argumentOIs;
+
+    transient Converter converter;
+
+    enum FUNC_NAMES {
+        ARRAY_MAX, ARRAY_MIN, ARRAY_DISTINCT, ARRAY_SLICE, ARRAY_JOIN, ARRAY_EXCEPT, ARRAY_INTERSECT
+    }
+
+    FUNC_NAMES FUNC_NAME;
+
+    @Override
+    public ObjectInspector initialize(ObjectInspector[] arguments)
+            throws UDFArgumentException {
+
+        // Check if wrong number of arguments were passed
+        checkArgsSize(arguments, MIN_ARG_COUNT, MAX_ARG_COUNT);
+
+        // Check if the argument is of category LIST or not
+        checkArgCategory(arguments, ARRAY_IDX, ObjectInspector.Category.LIST, FUNC_NAME,
+                org.apache.hadoop.hive.serde.serdeConstants.LIST_TYPE_NAME);
+
+        if (FUNC_NAME == FUNC_NAMES.ARRAY_EXCEPT
+                || FUNC_NAME == FUNC_NAMES.ARRAY_INTERSECT
+                || FUNC_NAME == FUNC_NAMES.ARRAY_JOIN) {
+            checkArgCategory(arguments, ARRAY2_IDX, ObjectInspector.Category.LIST, FUNC_NAME,
+                    org.apache.hadoop.hive.serde.serdeConstants.LIST_TYPE_NAME);
+        }
+
+        if (FUNC_NAME == FUNC_NAMES.ARRAY_SLICE) {
```

Comment:
Perhaps for this one, we can override the initialize, and check these categories in the derived class?